prov/efa/src/efa_cq.c

```diff
@@ -50,42 +49,53 @@ static uint64_t efa_cq_wc_to_fi_flags(struct efa_wc *wc)
 	}
 }
 
+/* @brief Unlike ibv_poll_cq, wide completion APIs do not write to ibv_wc struct. We need to do this manually.
+ * This is inspired by rdma-core, i.e. `efa_process_cqe` and `efa_process_ex_cqe`.
+ */
+static inline void efa_cq_cw_from_ibv_cq_ex_unsafe(struct ibv_cq_ex *cq, struct ibv_wc *wc) {
+	wc->status = cq->status;
+	wc->vendor_err = ibv_wc_read_vendor_err(cq);
+	wc->wc_flags = ibv_wc_read_wc_flags(cq);
+	wc->qp_num = ibv_wc_read_qp_num(cq);
+	wc->opcode = ibv_wc_read_opcode(cq);
+	wc->byte_len = ibv_wc_read_byte_len(cq);
+	wc->src_qp = ibv_wc_read_src_qp(cq);
+	wc->sl = ibv_wc_read_sl(cq);
+	wc->slid = ibv_wc_read_slid(cq);
+	wc->imm_data = ibv_wc_read_imm_data(cq);
+	wc->wr_id = cq->wr_id;
+}
+
 ssize_t efa_cq_readerr(struct fid_cq *cq_fid, struct fi_cq_err_entry *entry,
 		       uint64_t flags)
 {
 	struct efa_cq *cq;
-	struct efa_wce *wce;
-	struct slist_entry *slist_entry;
+	struct efa_wc *wc = {0};
 	uint32_t api_version;
 
 	cq = container_of(cq_fid, struct efa_cq, util_cq.cq_fid);
 
 	ofi_spin_lock(&cq->lock);
-	if (slist_empty(&cq->wcq))
-		goto err;
 
-	wce = container_of(cq->wcq.head, struct efa_wce, entry);
-	if (!wce->wc.ibv_wc.status)
+	efa_cq_cw_from_ibv_cq_ex_unsafe(cq->ibv_cq_ex, &wc->ibv_wc);
```

Comment:
at this point, I am not sure whether we need this step. Everything in `wc->ibv_wc` was from `cq->ibv_cq_ex`, then passed to the cq_err_entry. So why not just update `cq_err_entry` with `cq->ibv_cq_ex` directly?

Reply:
I'm ok either way in this case, since `efa_cq_readerr` is on the error path. I chose to do it here to make the code somewhat easier on the eye.

Will switch to `ibv_cq_ex` queries.